module/VuFindApi/src/VuFindApi/Controller/McpController.php

```diff
@@ -0,0 +1,155 @@
+     */
+    protected $baseAccessPermission = 'access.mcp';
+
+    /**
+     * JSON schema response code to represent an authorization error. This
+     * is not defined by any standard, except that the 32xxx range is app-specific.
+     *
+     * @var int
+     */
+    protected int $AUTH_ERROR = -32003;
+
+    /**
+     * Constructor
+     *
+     * @param ServiceLocatorInterface $sm     Service manager
+     * @param ?Server                 $server MCP Server
+     */
+    public function __construct(ServiceLocatorInterface $sm, protected ?Server $server)
+    {
+        parent::__construct($sm);
+    }
+
+    /**
+     * MCP action
+     *
+     * @return \Laminas\Http\Response
+     */
+    public function mcpAction()
+    {
+        if (!$this->server) {
+            throw new ServiceNotFoundException('This MCP server is not enabled.');
+        }
+
+        // $content = json_decode($this->params()->getController()->getRequest()->getContent(), true);
+        // $mcpMethod = $content['method'] ?? '';
+        $mcpMethod = null;
+        if ($this->isAccessDenied($mcpMethod)) {
+            // $messageId = $content['messageId'] ?? '';
+            $messageId = '';
+            return $this->outputAuthError($messageId);
+        }
+
+        // Adapting: https://github.com/modelcontextprotocol/php-sdk/blob/main/docs/transports.md
+        // and https://github.com/vufind-org/vufind/pull/4672/files#diff-89cf777c1454a4e7f97e51f800ca68001e874a555cb19ec27135779b76ccd8f4
+
+        // Convert to PSR-7 request
+        $psrRequest = ServerRequestFactory::fromGlobals();
+        foreach ($this->params()->fromRoute() as $routeParam => $value) {
+            $psrRequest = $psrRequest->withAttribute($routeParam, $value);
+        }
+
+        // Process with MCP
+        $transport = new StreamableHttpTransport($psrRequest);
+        $psrResponse = $this->server->run($transport);
+
+        // Convert back to Laminas response
+        if ($psrResponse instanceof ResponseInterface) {
+            return Psr7Response::toLaminas($psrResponse);
+        }
+        throw new \Exception('Unexpected state reached.');
+    }
+
+    /**
+     * Check whether access is denied based on the MCP method.
+     *
+     * @param string $mcpMethod MCP method
+     *
+     * @return bool
+     */
+    protected function isAccessDenied($mcpMethod): bool
+    {
+        $auth = $this->getService(\Lmc\Rbac\Mvc\Service\AuthorizationService::class);
+        return $mcpMethod
+            ? !$auth->isGranted($this->baseAccessPermission . '.' . $mcpMethod)
+            : !$auth->isGranted($this->baseAccessPermission);
+    }
+
+    /**
+     * Output an authorization error.
+     *
+     * @param string $messageId The MCP message Id.
+     *
+     * @return Response
+     */
+    protected function outputAuthError(string $messageId): Response
+    {
+        $error = new Error($messageId, $this->AUTH_ERROR, 'Access denied');
+        $response = $this->getResponse();
```

Comment:
Much of this is borrowed from ApiTrait.  Not sure if it's worth refactoring that method for the few bits that are useful here.